File: csharp/queen-attack/QueenAttack.cs
using System;

public class Queen
{
    public Queen(int row, int column)
    {
        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }
}

public static class QueenAttack
{
    public static bool CanAttack(Queen white, Queen black)
    {
        var dx = Math.Abs(white.Column - black.Column);
        var dy = Math.Abs(white.Row - black.Row);

        return dx == 0 || dy == 0 || dx == dy;
    }

    public static Queen Create(int row, int column)
    {
        if (row is < 0 or > 7 || column is < 0 or > 7)
        {
            throw new ArgumentOutOfRangeException();
        }

        return new Queen(row, column);
    }
}

File: julia/queen-attack/queen-attack.jl
struct InvalidPosition <: Exception
    row::Int
    col::Int
end
struct Queen
    row::Int
    col::Int
    Queen(row::Int, col::Int) = 0 ≤ row ≤ 7 && 0 ≤ col ≤ 7 ? new(row, col) : throw(InvalidPosition(row, col))
end

row(q::Queen) = q.row
column(q::Queen) = q.col

function canattack(white::Queen, black::Queen)
    dx, dy = abs(column(white) - column(black)), abs(row(white) - row(black))
    iszero(dx) || iszero(dy) || dx == dy
end
